torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1170,14 +1426,79 @@ def _register_param_handle(self, handle: FlatParamHandle) -> None:
         if handle not in self._handles:
             self._handles.append(handle)
 
+    @torch.no_grad()
+    def _unshard(
+        self,
+        handles: List[FlatParamHandle],
+    ) -> None:
+        """
+        Unshards the handles in ``handles``. If the handles are in
+        :meth:`summon_full_params` and are using mixed precision, then they are
+        forced to full precision.
+
+        Postcondition: Each handle's ``FlatParameter`` 's data is the padded
+        unsharded flattened parameter on the compute device.
+        """
+        any_ran_pre_unshard = False
+        with torch.cuda.stream(self._streams["pre_all_gather"]):
+            for handle in handles:
+                ran_pre_unshard = handle.pre_unshard()
```

Comment:
As mentioned in `flat_param.py`, `pre_unshard()` returns whether or not it was a no-op. This allows us to only call `wait_stream()` if needed.